aten/src/ATen/native/mkl/SpectralOps.cpp

```diff
@@ -0,0 +1,297 @@
+          // Substract. Carries over to previous dimension
+          from_slice_data -= output.stride(d);
+        } else if (from_slice_indices[d] == 1) {
+          // Dimension index becomes 1
+          // Doesn't carry over to previous dimension
+          from_slice_data += (output.size(d) - 1) * output.stride(d);
+          break;
+        } else {
+          // Substract. Doesn't carry over to previous dimension
+          from_slice_data -= output.stride(d);
+          break;
+        }
+      } else {
+        // At d = 0 nonbatch dim, it means that to_slice_data ise now at a the
+        // beginning of a data sample. It maps to itself by conjugate symmetry.
+        from_slice_data = to_slice_data;
+      }
+    }
+    num--;
+  }
+}
+
+// input should be a contiguous batched tensor of same size as full (twosided)
+// signals, but only contains half (onesided) of the values.
+// This function modifies inplace.
+static inline void _fft_fill_with_conjugate_symmetry_(Tensor& input,
+                      int64_t signal_ndim, int64_t size_last_dim,
+                      int64_t last_dim_start_slice) {
+  if (last_dim_start_slice >= size_last_dim) {
+    return;
+  }
+
+  int64_t num = 1;
+  for (int64_t d = 0; d < signal_ndim; d++) {
+    num *= input.size(d);
+  }
+#ifdef _OPENMP
+  if (num > 500) {
+    int nthreads = omp_get_num_threads();
+    int64_t num_slices_per_thread = num / nthreads + 1;
+    #pragma omp parallel
+    {
+      int tid = omp_get_thread_num();
+      int64_t start = tid * num_slices_per_thread;
+      AT_DISPATCH_FLOATING_TYPES(input.type(), "_fft_fill_with_conjugate_symmetry", [&] {
+        _fft_fill_with_conjugate_symmetry_slice<scalar_t>(input, signal_ndim, size_last_dim,
+            last_dim_start_slice, start, std::min(num_slices_per_thread, num - start));
+      });
+    }
+    return;
+  }
+#endif
+  AT_DISPATCH_FLOATING_TYPES(input.type(), "_fft_fill_with_conjugate_symmetry", [&] {
+    _fft_fill_with_conjugate_symmetry_slice<scalar_t>(input, signal_ndim, size_last_dim,
+        last_dim_start_slice, 0, num);
+  });
+}
+
+// Returns undefined tensor if the parameter range is greater than MKL_LONG.
+Tensor _fft_mkl(const Tensor& self, int64_t signal_ndim,
+                bool complex_input, bool complex_output,
+                bool inverse, IntList checked_signal_sizes,
+                bool normalized, bool onesided,
+                IntList output_sizes) {
+  int64_t batch = self.size(0);
+  Tensor input = self;
+  // real/imag dimension must be like complex type
+  if (complex_input && input.stride(-1) != 1) {
+    input = input.contiguous();
+  }
+
+  // check if we can use MKL because MKL_LONG is 32bit on some OS, e.g. Windows
+  // need to check input and output size and strides
+  // be careful about complex domain, where the stride needs to be divided by 2
+  // only need to test upper bound MKL_LONG_MAX as these values are non-negative
+  if (sizeof(MKL_LONG) < sizeof(int64_t)) {
```

Comment:
Where are the mkl contiguity requirements documented?

Reply:
MKL is nicer than cuFFT. It takes in strides directly. So non-contiguous tensors are fine, as long as we can represent everything using `MKL_LONG`.